File: src/main/java/ideaclock/ClockStatusBarWidget.java
package ideaclock;

import com.intellij.openapi.wm.StatusBar;
import com.intellij.openapi.wm.StatusBarWidget;
import com.intellij.util.Consumer;
import com.intellij.util.text.DateFormatUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.MouseEvent;

public class ClockStatusBarWidget implements StatusBarWidget, StatusBarWidget.TextPresentation, ActionListener {

    private static final long MINUTES = 60000;
    private final Timer timer = new Timer(1000, this);
    private StatusBar statusBar;

    private volatile long lastTime = getTime();

    @NotNull
    @Override
    public String ID() {
        return "IdeaClock";
    }

    @Nullable
    @Override
    public WidgetPresentation getPresentation(@NotNull PlatformType platformType) {
        return this;
    }


    @Override
    public void install(@NotNull StatusBar statusBar) {
        this.statusBar = statusBar;
        timer.start();

    }

    @Override
    public void dispose() {
        timer.stop();
        statusBar = null;
    }

    @Override
    public void actionPerformed(ActionEvent actionEvent) {
        long newTime = getTime();
        if (newTime / MINUTES != lastTime / MINUTES) {
            lastTime = newTime;
            statusBar.updateWidget(ID());
        }
        timer.restart();
    }

    @NotNull
    @Override
    public String getText() {
        return "It's: " + DateFormatUtil.formatTime(lastTime);
    }

    @NotNull
    @Override
    public String getMaxPossibleText() {
        return null;
    }

    @Override
    public float getAlignment() {
        return 0.5f;
    }

    @Nullable
    @Override
    public String getTooltipText() {
        return "time...";
    }

    @Nullable
    @Override
    public Consumer<MouseEvent> getClickConsumer() {
        return null;
    }

    @NotNull
    long getTime() {
        return System.currentTimeMillis();
    }
}

File: src/main/kotlin/ideaclock/ClockStatusBarWidget.kt
package ideaclock

import com.intellij.openapi.wm.StatusBar
import com.intellij.openapi.wm.StatusBarWidget
import com.intellij.util.Consumer
import com.intellij.util.text.DateFormatUtil

import javax.swing.*
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.MouseEvent

class ClockStatusBarWidget : StatusBarWidget, StatusBarWidget.TextPresentation, ActionListener {
    private val timer = Timer(1000, this)
    private var statusBar: StatusBar? = null

    @Volatile private var lastTime = time

    override fun ID(): String {
        return "IdeaClock"
    }

    override fun getPresentation(platformType: StatusBarWidget.PlatformType): StatusBarWidget.WidgetPresentation? {
        return this
    }


    override fun install(statusBar: StatusBar) {
        this.statusBar = statusBar
        timer.start()

    }

    override fun dispose() {
        timer.stop()
        statusBar = null
    }

    override fun actionPerformed(actionEvent: ActionEvent) {
        val newTime = time
        if (newTime / MINUTES != lastTime / MINUTES) {
            lastTime = newTime
            statusBar?.updateWidget(ID())
        }
        timer.restart()
    }

    override fun getText(): String {
        return "\u25F7 " + DateFormatUtil.formatTime(lastTime)
    }

    override fun getMaxPossibleText(): String {
        return "0000000"
    }

    override fun getAlignment(): Float {
        return 0.5f
    }

    override fun getTooltipText(): String? {
        return "time..."
    }

    override fun getClickConsumer(): Consumer<MouseEvent>? {
        return null
    }

    internal val time: Long
        get() = System.currentTimeMillis()

    companion object {
        private val MINUTES: Long = 60000
    }
}
